src/solvers/dgsem_p4est/containers_2d_manifold_in_3d_covariant.jl

```diff
@@ -0,0 +1,209 @@
+function compute_auxiliary_variables!(elements, mesh::P4estMesh{2, 3},
+                                      equations::AbstractCovariantEquations{2, 3}, dg)
+
+    # The tree node coordinates are assumed to be on the spherical shell centred around the 
+    # origin. Therefore, we can compute the radius once and use it throughout.
+    radius = sqrt(mesh.tree_node_coordinates[1, 1, 1, 1]^2 +
+                  mesh.tree_node_coordinates[2, 1, 1, 1]^2 +
+                  mesh.tree_node_coordinates[3, 1, 1, 1]^2)
+
+    for element in 1:Trixi.ncells(mesh)
+
+        # Check that the degree of the mesh matches that of the solver
+        @assert length(mesh.nodes) == nnodes(dg)
+
+        # Extract the corner vertex positions from the P4estMesh
+        v1 = Trixi.get_node_coords(mesh.tree_node_coordinates, equations, dg,
+                                   1, 1, element)
+        v2 = Trixi.get_node_coords(mesh.tree_node_coordinates, equations, dg,
+                                   nnodes(dg), 1, element)
+        v3 = Trixi.get_node_coords(mesh.tree_node_coordinates, equations, dg,
+                                   nnodes(dg), nnodes(dg), element)
+        v4 = Trixi.get_node_coords(mesh.tree_node_coordinates, equations, dg,
+                                   1, nnodes(dg), element)
+
+        for j in eachnode(dg), i in eachnode(dg)
+
+            # get reference node coordinates
+            xi1, xi2 = dg.basis.nodes[i], dg.basis.nodes[j]
+
+            # Construct a bilinear mapping based on the four corner vertices
+            x_bilinear = 0.25f0 *
+                         ((1 - xi1) * (1 - xi2) * v1 + (1 + xi1) * (1 - xi2) * v2 +
+                          (1 + xi1) * (1 + xi2) * v3 + (1 - xi1) * (1 + xi2) * v4)
+
+            # Project the mapped local coordinates onto the sphere 
+            scaling_factor = radius / norm(x_bilinear)
+            x = scaling_factor * x_bilinear
+
+            # Convert to longitude and latitude
+            lon, lat = atan(x[2], x[1]), asin(x[3] / radius)
+
+            # Compute trigonometric terms needed for analytical metrics
+            sinlon, coslon = sincos(lon)
+            sinlat, coslat = sincos(lat)
+            a11 = sinlon * sinlon * coslat * coslat + sinlat * sinlat
+            a12 = -sinlon * coslon * coslat * coslat
+            a13 = -coslon * sinlat * coslat
+            a21 = a12
+            a22 = coslon * coslon * coslat * coslat + sinlat * sinlat
+            a23 = -sinlon * sinlat * coslat
+            a31 = -coslon * sinlat
+            a32 = -sinlon * sinlat
+            a33 = coslat
+
+            # Analytically compute the transformation matrix A, such that G = AᵀA is the 
+            # covariant metric tensor and a_i = A[1,i] * e_lon + A[2,i] * e_lat denotes 
+            # the covariant tangent basis, where e_lon and e_lat are the unit basis vectors
+            # in the longitudinal and latitudinal directions, respectively.
```

Comment:
Does the use of `e_lon` and `e_lat` imply a problem in (the vicinity of) the pole singularities? if it does, it should be properly stated. Moreover, is this the standard way to do this? Wouldn't it be a bit more general to use Cartesian unit vectors to store the covariant basis vectors?

Reply:
This is the standard way of doing things, and does not introduce any pole problem. This matrix is, in my opinion at least, more convenient than one using Cartesian coordinates, because it converts to/from spherical velocity components, which are typically more meaningful than Cartesian ones in a geophysical context (e.g. initial conditions are commonly defined in terms of zonal and meridional components, not Cartesian ones). It also is 2x2 instead of 3x2 and thus (marginally) cheaper for 2D manifolds.

Reply:
You are right, the initial conditions are commonly defined in terms of zonal and meridional components. However my concern is the following:

When I run the code above with
```julia
v1 = [0.0, 0.0, 1.0]
v2 = [1.0, 0.0, 0.0]
v3 = [0.0, 1.0, 0.0]
v4 = [0.0, sqrt(2.0) * 0.5, sqrt(2.0) * 0.5]
xi1, xi2 = -1.0, -1.0
```
then I get the covariant basis
```julia
julia> basis_covariant
2×2 SMatrix{2, 2, Float64, 4} with indices SOneTo(2)×SOneTo(2):
  0.0   0.353553
 -0.5  -8.96727e-18
```
However, if I run the code with
```julia
v1 = [eps(1.0), eps(1.0), 1.0]
v2 = [1.0, 0.0, 0.0]
v3 = [0.0, 1.0, 0.0]
v4 = [0.0, sqrt(2.0) * 0.5, sqrt(2.0) * 0.5]
xi1, xi2 = -1.0, -1.0
```
then I get the matrix
```julia
julia> basis_covariant
2×2 SMatrix{2, 2, Float64, 4} with indices SOneTo(2)×SOneTo(2):
 -0.353553   0.25
 -0.353553  -0.25
```
Very different(!!). Isn't that problematic?... This transformation matrix is used very often to transform the state quantities at the interfaces. What happens if some interface nodes are directly at / in the vicinity of the pole singularity?

Reply:
If we let $L$, $R$, and $S$ denote the left, right, and global coordinate systems, what ends up happening with the transformation matrices is that you have 
```math
(v^1, v^2)_R = [A_{L \to R}]\, (v^1, v^2)_L
```
where 
```math
  A_{L \to R} = [A_{R\to G}]^{-1} [A_{L \to G}].
```
If the matrices $A_{R\to G}$ and $A_{L \to G}$ are not (nearly) singular, does it actually matter that there is a pole?

Reply:
The risk, in my opinion, is that the global coordinate system is differently defined for two elements at different sides of the interface due to machine precision errors.

Reply:
I see. Could it be that even if $A_{L \to R}$ and  $A_{R \to L}$ are inverses of each other up to roundoff (this is what we need for conservation) there still may be excessive error introduced that doesn't show up as conservation error? I presume this would be a problem in other codes too.

Reply:
I wonder if we should make a separate PR for improvements to this pole issue, as in my opinion, this is more of a scientific problem to be investigated, rather than an implementation error.

Reply:
I opened the issue https://github.com/trixi-framework/TrixiAtmo.jl/issues/49
I suggest that we merge this PR but leave this conversation "unresolved" for the moment.